src/librustc/infer/error_reporting/need_type_info.rs

```diff
@@ -187,39 +291,36 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             //          ^ consider giving this closure parameter the type `[_; 0]`
             //            with the type parameter `_` specified
             // ```
-            labels.clear();
-            labels.push((
+            err.span_label(
                 pattern.span,
                 format!("consider giving this closure parameter {}", suffix),
-            ));
+            );
         } else if let Some(pattern) = local_visitor.found_local_pattern {
             if let Some(simple_ident) = pattern.simple_ident() {
```

Comment:
```suggestion
            let msg = if let Some(simple_ident) = pattern.simple_ident() {
```
`err.span_label(pattern.span, msg)` is in common for the `if let` + `match` and the `else`.